pyannote/audio/core/pipeline.py

```diff
@@ -187,6 +182,133 @@ def from_pretrained(
+                when running `huggingface-cli login` (stored in `~/.huggingface`). Will default to `True` if `repo_url`
+                is not specified.
+            create_pr (`bool`, *optional*, defaults to `False`):
+                Whether or not to create a PR with the uploaded files or directly commit.
+            revision (`str`, *optional*):
+                Branch to push the uploaded files to.
+            commit_description (`str`, *optional*):
+                The description of the commit that will be created
+            tags (`List[str]`, *optional*):
+                List of tags to push on the Hub.
+            cache_dir: Path or str, optional
+                Path to config cache directory. Defauorch/pyannote" when unset.
+            config_yaml_path (`str`):
+                Path to the original "pyannote/speaker-diarization-3.1" Hub repository containing the default config file to use
+                to create custom speaker diarization config files.
+        """
+
+        api = HfApi()
+
+        _ = api.create_repo(
+            repo_id, private=private, token=token, exist_ok=True, repo_type="model"
+        )
+
+        # Load the pyannote/speaker-diarization-3.1 pipeline config:
+        try:
+            config_yml = hf_hub_download(
+                config_yaml_path,
+                "config.yaml",
+                repo_type="model",
+                revision=revision,
+                library_name="pyannote",
+                library_version=__version__,
+                cache_dir=cache_dir,
+                use_auth_token=token,
+            )
+        except RepositoryNotFoundError:
+            print(
+                f"""
+                Could not download '{config_yaml_path}' pipeline.
+                It might be because the pipeline is private or gated so make
+                sure to authenticate. Visit https://hf.co/settings/tokens to
+                create your access token and retry with:
+
+                >>> Pipeline.from_pretrained('{config_yaml_path}',
+                ...                          use_auth_token=YOUR_AUTH_TOKEN)
+
+                If this still does not work, it might be because the pipeline is gated:
+                visit https://hf.co/{config_yaml_path} to accept the user conditions."""
+            )
+            return None
+
+        with TemporaryDirectory() as tmpdir:
+
+            tmpdir = Path(tmpdir)
+
+            with open(config_yml, "r") as fp:
+                config = yaml.load(fp, Loader=yaml.SafeLoader)
+
+            if embedding_model is None:
+                embedding_model = self.embedding
+
+            if segmentation_model is None:
+                segmentation_model = self.segmentation_model
+
+            # Modify the config with new segmentation and embedding models:
+            config["pipeline"]["params"]["embedding"] = embedding_model
```

Comment:
As discussed offline: an elegant solution would be to save both the embedding and segmentation models to subfloders in the repo (`embedding` and `segmentation` respectively), and then load the weights from these subfolders when we call `.from_pretrained`

Your repo structure for `kamilakesbi/spd_pipeline_test` could look something like the following:
```markdown
├── config.yaml                    <- Top-level pipeline config
├── embedding                 <- Subfolder for the embedding model
|   ├── config.yaml             
|   ├── pytorch_model.bin
├── segmentation                 <- Subfolder for the segmentation model
|   ├── config.yaml             
|   ├── pytorch_model.bin
```

And your top-level yaml file could have an extra entry:

```yaml
    embedding: kamilakesbi/spd_pipeline_test
    embedding_subfolder: embedding
    ...
    segmentation: kamilakesbi/spd_pipeline_test
    segmentation_subfolder: segmentation
```

Note that this would require updating `.from_pretrained` to handle this extra subfolder logic

Reply:
Like it! 

Here, what would be the role of `embedding` and `segmentation` if we have `embedding_subfolder ` and `segmentation_subfolder ` ? 

`embedding` and `segmentation` are currently pointers to checkpoints on the hub. If we add the subfolders, we won't need these pointers anymore! 

--> One option could be to make subfolders arguments  optional (or `embedding` and `segmentation` optionnal).

Reply:
I handled this differently: 

- If we want to save the checkpoints, we add a `save_checkpoints=True` parameter to `pipeline.push_to_hub`.  We would then get a repo structure like the one you proposed @sanchit-gandhi, but the top yaml file would look like this: 

```diff
checkpoints: True
params: 
- embedding: 'kamilakesbi/speaker-embedding-test'
- embedding: 'kamilakesbi/speaker-segmentation-test'
```

- If we don't want to store checkpoints on the hub, then we need pointers to the segmentation and embedding models on the hub. In this case the config file would look like this: 

 ```diff
checkpoints: False
params: 
+ embedding: 'kamilakesbi/speaker-embedding-test'
+ embedding: 'kamilakesbi/speaker-segmentation-test'
```

Reply:
As discussed offline, saving all sub-models means the model repo on the Hub is fully portable -> users can clone the repository and have all sub-models available to them locally

This is the design that was adopted for [diffusers pipelines](https://huggingface.co/docs/diffusers/v0.27.2/en/using-diffusers/loading#diffusionpipeline-explained) and has worked very well

Thus, we'll assume that any new checkpoints being pushed will follow this new repo structure, with an exception for current pipelines on the Hub that leverage components from multiple repositories